README.md
SIGVerse/SIGViewer
=============
See [CHANGES.md](CHANGES.md) for an overview of the differences between this fork and the original [SIGVerse SIGViewer](https://github.com/SIGVerse/SIGViewer).


### To Enable Oculus Rift Mode

open: `SIGVerse.ini`

And then change `OCULUS_MODE=false` to `OCULUS_MODE=true`

## Dependencies
To build this project and its dependecies, you need:

* [SIGService](https://github.com/noirb/SIGService/tree/dev)
* [X3D Parser](https://github.com/noirb/x3d/tree/dev)
* [The Ogre SDK](http://www.ogre3d.org/download/sdk), version 1.9
* [GLEW](http://glew.sourceforge.net/)
* [CEGUI](http://cegui.org.uk/), version 0.8+ (built with Ogre support)
* [Java JDK](http://www.oracle.com/technetwork/java/javase/downloads/index.html)
* [Boost](http://www.boost.org/users/history/version_1_61_0.html), version 1.61.0+
* [LibSSH2](https://www.libssh2.org/)
* [OpenSSL](https://www.openssl.org/)
* [Zlib](http://www.zlib.net/) (actually included with CEGUI's dependencies, so not necessary to install separately)
* [Oculus SDK](https://developer.oculus.com/downloads/), version 1.6.0
* [CMake](https://cmake.org/download/)
* (recommended) [7-Zip](http://www.7-zip.org/)

A PowerShell script, `setup_env.ps1`, is provided to help build and configure these in order to get you a working development environment as quickly as possible.
Run this script from the directory you want to make your "SIGVerse Root" (e.g. one directory level above the location you cloned SIGViewer to).
It will:

1. Detect any Visual Studio installations on your machine (and prompt you to choose one if there are several)
2. Check for CMake and 7-Zip, and if they can't be found prompt you for their location (in case you have them on your desktop or something)
3. Download recent versions of all of the above dependencies, excluding [JDK](http://www.oracle.com/technetwork/java/javase/downloads/index.html), [CMake](https://cmake.org/download/) and [7-Zip](http://www.7-zip.org/) (which you should install yourself before starting). If you're using a version of Visual Studio newer than 2012, you will also need to [download the Ogre SDK yourself from here](http://ogre3d.org/forums/viewtopic.php?t=69274), since those builds aren't hosted in a location the script can automatically pull from.
4. Unzip them into your "SIGVerse Root" with the following structure:
    `SIGVerseRoot\*        -- SIGVerse projects (SIGViewer, etc.)`
    `SIGVerseRoot\external -- All other dependencies`
5. Compile necessary Boost libraries used by Ogre and CEGUI
6. Compile CEGUI and its dependencies
7. Compile libSSH2 (using the zlib from CEGUI's dependencies)
8. Compile X3D and SIGService
9. Generate a script named `setenv.bat`, which can be invoked any time in the future to set all the necessary environment variables to build SIGViewer. After running this script, you can open the project from the commandline via `devenv SIGViewer_2010.sln`.

Before running the powershell script, you may need to change powershell's execution policy, which is `restricted` by default. To do so, run powershell as an administrator, and then execute the command: `Set-ExecutionPolicy RemoteSigned`

This will allow you to run local scripts, even when *not* running powershell as an administrator (which is not required for any of the scripts in this project, and is in general not recommended unless you really *need* it).

It should only be necessary to run `setup_env.ps1` once, but if you do run it again it will attempt to detect any existing code and prompt you for what you want to keep and what you want to replace (in this way you can update just one library without having to re-download and rebuild everything).

## Building
**Always make sure to use the same compiler and release settings across dependencies where applicable! This means if you're building the 32-bit Release version of SIGViewer, you should be building the 32-bit Release version of all its dependencies as well!**

The `setup_env.ps1` script will generate 32-bit Release builds of all dependecies (if there's demand, it will be updated to allow Debug builds as well, but SIGViewer does not currently support 64-bit builds). To use it, simply run the script from the location you want all of your SIGVerse-related files to live. This is most likely the directory in which you put the SIGViewer code (e.g. if SIGViewer is in `C:\SIGVerse\SIGViewer`, then `C:\SIGVerse` is the `SIGVerse Root` you want to run the script from).

The script will first check to ensure Visual Studio, CMake, and 7-zip are installed, then allow you to confirm that the correct versions of each are selected before beginning. It will then download and set up the necessary dependecies, and give you one last chance to confirm or cancel before building them all. If any errors are detected they should be logged in the script output.

```
    C:\Users\nao\SIGVerse> .\SIGViewer\setup_env.ps1 
    Found the following Visual Studio installations:
    [ 1 ] : Visual Studio 14 2015
    Selected: Visual Studio 14 2015
    --
    Please download the OGRE SDK for your compiler from here: http://ogre3d.org/forums/viewtopic.php?t=69274
    And extract it into the directory: C:\Users\nao\SIGVerse\extern\OGRE-SDK-1.9.0-vc140-x86-12.03.2016
    --
    Press enter to continue...:

    7-Zip not found in PATH. Searching for installation directory.

    Current working directory: C:\Users\nao\SIGVerse

CHANGES.md
# noirb/SIGViewer

This document outlines the differences between this fork of SIGViewer and the [original SIGViewer](https://github.com/SIGVerse/SIGViewer).

## Overview

Briefly, the following changes have been made:

* Support for Oculus Rift CV1 (up to SDK v. 1.6.0)
* Streamlined setup and build process; support for multiple versions of Visual Studio
* Updated to more recent and better-supported version of some libraries

## Details

### Project Setup and Build Process

#### Project Setup

Due to the large number of dependencies the project has, and the need to ensure that many of them are compiled together with the same settings, a setup script, `setup_env.ps1`, to automatically download and configure as many dependencies as possible was written.

Relevant Commits:

[e6c0f21f21bbd38828ae2d1a66f89430fce6812c](https://github.com/noirb/SIGViewer/commit/e6c0f21f21bbd38828ae2d1a66f89430fce6812c)
[ae23891e7cde9a9841397ac5de17ced8533da903](https://github.com/noirb/SIGViewer/commit/ae23891e7cde9a9841397ac5de17ced8533da903)
[f20731adfae56888e6bd2a0784b56613e2a50c17](https://github.com/noirb/SIGViewer/commit/f20731adfae56888e6bd2a0784b56613e2a50c17)
[5c1cca4dcb3103010591f7e45022cc5cec754bad](https://github.com/noirb/SIGViewer/commit/5c1cca4dcb3103010591f7e45022cc5cec754bad)
[d00362eeebeafb6a169bc027e5fd4328d6ca7801](https://github.com/noirb/SIGViewer/commit/d00362eeebeafb6a169bc027e5fd4328d6ca7801)
[62bb212daef592921c666308859e6ce61adac9ae](https://github.com/noirb/SIGViewer/commit/62bb212daef592921c666308859e6ce61adac9ae)

#### Building

To make the build process a bit more flexible and to make debugging build issues easier, the project was configured to reference the necessary headers and libraries via environment variables (which are automatically set by the script mentioned above) rather than with hard-coded paths. This allows users to customize their development environment more easily.

Similar changes were also made in forks of [SIGService](https://github.com/noirb/SIGService) and [X3D](https://github.com/noirb/x3d), so all the relevant SIGVerse projects can be set up and built from the same console.

Relevant Commits:

[87e230ec738a142b762921a7eff2470b14ee8acd](https://github.com/noirb/SIGViewer/commit/87e230ec738a142b762921a7eff2470b14ee8acd)
[32fbb54b4478d8b369cb0856685ba6646eab8e37](https://github.com/noirb/SIGViewer/commit/32fbb54b4478d8b369cb0856685ba6646eab8e37)
[3dd9beec29bfd892cc85f5ce2ed3c47719696ce4](https://github.com/noirb/SIGViewer/commit/3dd9beec29bfd892cc85f5ce2ed3c47719696ce4)

### CEGUI 0.8+


setup_env.ps1

$scriptPath = split-path -parent $MyInvocation.MyCommand.Definition
# The root directory of the whole project
$projectRoot = $PWD.Path

# A directory to store temporary downloaded files
$tmp_dir = "$projectRoot\setup_tmp"
if ( !(Test-Path $tmp_dir) ) {
    new-item $tmp_dir -type directory
}

# The root directory to store non-SIGVerse dependencies in
$projectDepsRoot = "$projectRoot\extern"
if ( !(Test-Path $projectDepsRoot) ) {
    new-item $projectDepsRoot -type directory
}

# environment variables used to build the project
$build_vars = @{
    "OGRE_SDK"             = "";
    "GLEW_ROOT_PATH"       = "";
    "CEGUI_ROOT_PATH"      = "";
    "CEGUI_DEPS_ROOT"      = "";
    "LIBSSH2_ROOT_PATH"    = "";
    "OPENSSL_ROOT_DIR"     = "";
    "X3D_ROOT_PATH"        = "";
    "SIGSERVICE_ROOT_PATH" = "";
    "BOOST_ROOT"           = "";
    "LIBOVR_ROOT_PATH"     = "";
    "VS_TOOLS_PATH"        = "";
    "OPENCV_ROOT"          = "";
}

# These are used as an argument to CMake when generating project files
$vsVersionNames = @{
    "VS14" = "Visual Studio 14 2015";
    "VS12" = "Visual Studio 12 2013";
    "VS11" = "Visual Studio 11 2012";
    "VS10" = "Visual Studio 10 2010";
    "VS09" = "Visual Studio 9 2008"
}

# Find installed visual studio versions
$vsEnvVars = @()

foreach ($item in (dir Env:)) {
	if ($item.name -Match "VS[0-9]{1,3}COMNTOOLS") {
		$vsEnvVars = $vsEnvVars + $item
	}
}

# If no VS installation was found, bail
if ($vsEnvVars.length -eq 0) {
	echo 'No Version of Visual Studio was detected! Please ensure you have VS 2010 or later installed!'
	echo 'Exiting...'
	exit
}

echo 'Found the following Visual Studio installations:'
$iter = 1
foreach ($item in $vsEnvVars) {
    $shortName = $item.Key.Substring(0,4)
    if ($vsVersionNames.ContainsKey($shortName))
    {
        write-host [ $iter ] : $vsVersionNames[$shortName]
    }
    else
    {
        write-host [ $iter ] : Unknown VS Version: $iter.Key
    }
    $iter++
}

$vsVersion = ""
$vsToolsPath = ""

if ($vsEnvVars.length -gt 1) {

    do{
        $resp = Read-Host -prompt "Select Version (1-$($iter-1))"
        [int]$choice = $null
        [void][int32]::TryParse( $resp, [ref]$choice )
        if ( !$choice -Or $choice -lt 1 -Or $choice -ge $iter)
        {
          write-host "Please enter a number between 1 and $($iter-1)"
        }
        else
        {
            $vsVersion   = $vsVersionnames[$vsEnvVars[$choice-1].Key.Substring(0,4)]
            $vsToolsPath = $vsEnvVars[$choice-1].Value
            write-host "Selected: $vsVersion"
        }
      } until ($vsVersion.length -gt 0 -And $vsToolsPath.length -gt 0)
}
else {
    $vsVersion   = $vsVersionNames[$vsEnvVars[0].Key.Substring(0,4)]
    $vsToolsPath = $vsEnvVars[0].Value
    write-host "Selected: $vsVersion"
}

# if we're using VS 2015 or 2013, OGRE SDK must be manually downloaded...
if ($vsVersion -eq "Visual Studio 14 2015")
{
    write-host "--"
    write-host "Please download the OGRE SDK for your compiler from here: http://ogre3d.org/forums/viewtopic.php?t=69274"
    write-host "And extract it into the directory: $projectDepsRoot\OGRE-SDK-1.9.0-vc140-x86-12.03.2016"
    write-host "--"
    Read-Host "Press enter to continue..."
    $build_vars.Set_Item("OGRE_SDK", "$projectDepsRoot\OGRE-SDK-1.9.0-vc140-x86-12.03.2016")
}
if ($vsVersion -eq "Visual Studio 12 2013")
{
    write-host "--"
    write-host "Please download the OGRE SDK for your compiler from here: http://ogre3d.org/forums/viewtopic.php?t=69274"
    write-host "And extract it into the directory: $projectDepsRoot\OGRE-SDK-1.9.0-vc120-x86-12.03.2016"
    write-host "--"
    Read-Host "Press enter to continue..."
    $build_vars.Set_Item("OGRE_SDK", "$projectDepsRoot\OGRE-SDK-1.9.0-vc120-x86-12.03.2016")
}

# Find 7-Zip
$7zPath = ''
foreach ($path in (($Env:path).split(";"))) {
    if ($path -like "*7-Zip*" ) {
        $7zPath = $path
    }
}
# if it wasn't in the PATH, go hunting in program files...
if (!$7zPath) {
    echo "7-Zip not found in PATH. Searching for installation directory."
    if (Test-Path "$Env:ProgramFiles\7-Zip") {
        $7zPath = "$Env:ProgramFiles\7-Zip"
    }
    elseif (Test-Path "${Env:ProgramFiles(x86)}\7-Zip") {
        $7zPath = "${Env:ProgramFiles(x86)}\7-Zip"
    }
    # if it wasn't found, ask the user for a path in case they have a portable copy
    else {
        echo "ERROR: Could not find 7-Zip!"
        echo "Please enter the path to your 7-Zip directory (ctrl+C to cancel)"
        $7zPath = Read-Host -prompt "7-Zip Path"
    }
}
$7zX = "& ""$7zPath\7z.exe"" x -y"

# Find CMake
$cmakePath = ''
foreach ($path in (($Env:path).split(";"))) {
    if ($path -like "*CMake*" ) {
        $cmakePath = $path
    }
}
if (!$cmakePath) {
    echo "CMake not found in PATH. Searching for installation directory."
    foreach ($path in (dir ${Env:ProgramFiles(x86)})) {
        if ($path -like "*CMake*") {
            $cmakePath = "${Env:ProgramFiles(x86)}\$path\bin"
        }
    }
    foreach ($path in (dir $Env:ProgramFiles)) {
        if ($path -like "*CMake*") {
            $cmakePath = "$($Env:ProgramFiles)\$path\bin"
        }
    }
}
if (!$cmakePath) {
    echo "ERROR: Could not find CMake!"
    echo "Please enter the path to your CMake installation directory (ctrl+C to cancel)"
    $cmakePath = Read-Host -prompt "CMake Path"
}
$cmake = """$cmakePath\cmake.exe"""


$build_vars.Set_Item("VS_TOOLS_PATH", $vsToolsPath)

echo ""
echo "Current working directory: $projectRoot"
echo "========================================================="
echo "    Temporary files will be copied to:   $tmp_dir"
echo "    SIGVerse projects will be set up in: $projectRoot"
echo "    Dependencies will be set up in:      $projectDepsRoot"
echo "    Visual Studio:"
echo "                  Version: $vsVersion"
echo "                  Tools:   $($build_vars.Get_Item(""VS_TOOLS_PATH""))"
echo "    7-Zip Directory:                     $7zPath"
echo "    CMake Directory:                     $cmakePath"
echo "========================================================="

if ( !((Read-Host -Prompt "Proceed with this configuration? (y/n)").ToLower().StartsWith("y")) ) {
    exit
}

# Locations to download dependencies from
## TODO: check for git and prefer cloning over downloading when possible
$SIGViewer_www  = "https://github.com/noirb/SIGViewer/archive/dev.zip"
$SIGService_www = "https://github.com/noirb/SIGService/archive/dev.zip"
$X3D_www        = "https://github.com/noirb/x3d/archive/dev.zip"
$CEGUI_www      = "http://prdownloads.sourceforge.net/crayzedsgui/cegui-0.8.7.zip"
$CEGUI_deps_www = "http://prdownloads.sourceforge.net/crayzedsgui/cegui-deps-0.8.x-src.zip"
$libSSH2_www    = "https://www.libssh2.org/download/libssh2-1.7.0.tar.gz"
$openSSL_www    = "https://openssl-for-windows.googlecode.com/files/openssl-0.9.8k_WIN32.zip"
$boost_www      = "http://downloads.sourceforge.net/project/boost/boost/1.61.0/boost_1_61_0.zip"
$libovr_www     = "https://static.oculus.com/sdk-downloads/0.8.0.0/Public/1445451746/ovr_sdk_win_0.8.0.0.zip"
$glew_www       = "http://downloads.sourceforge.net/project/glew/glew/2.0.0/glew-2.0.0-win32.zip"
$opencv_www     = "http://downloads.sourceforge.net/project/opencvlibrary/opencv-win/2.4.13/opencv-2.4.13.exe"

# set download URL for Ogre only if we're using VS 2010 or 2012
if ($vsVersion -eq "Visual Studio 11 2012")
{
	$Ogre_SDK_www   = "http://downloads.sourceforge.net/project/ogre/ogre/1.9/1.9/OgreSDK_vc11_v1-9-0.exe"
}
if ($vsVersion -eq "Visual Studio 10 2010")
{
	$Ogre_SDK_www   = "http://downloads.sourceforge.net/project/ogre/ogre/1.9/1.9/OgreSDK_vc10_v1-9-0.exe"
}

$net = new-object System.Net.WebClient

function doDownload
{
    param( [string]$url, [string]$destDir, [string]$destFile, [System.Net.WebClient]$wc )
    
    if ( Test-Path "$destDir\$destFile" ) {
        $conf = Read-Host -Prompt "An existing $destFile was found! Download a fresh copy? (y/n)"
        if ( $conf.ToLower().StartsWith("y") ) {
            echo "Downloading $destFile..."
            $wc.DownloadFile($url, "$destDir\$destFile");
        }
    }
    else {
        echo "Downloading $destFile..."
        $wc.DownloadFile($url, "$destDir\$destFile");
    }
}

# download all 
doDownload -url $SIGViewer_www  -destFile "sigviewer.zip"  -destDir $tmp_dir -wc $net
doDownload -url $SIGService_www -destFile "sigservice.zip" -destDir $tmp_dir -wc $net
doDownload -url $X3D_www        -destFile "x3d.zip"        -destDir $tmp_dir -wc $net
doDownload -url $CEGUI_www      -destFile $CEGUI_www.Substring($CEGUI_www.LastIndexOf("/") + 1)           -destDir $tmp_dir -wc $net
doDownload -url $CEGUI_deps_www -destFile $CEGUI_deps_www.Substring($CEGUI_deps_www.LastIndexOf("/") + 1) -destDir $tmp_dir -wc $net
doDownload -url $libSSH2_www    -destFile $libSSH2_www.Substring($libSSH2_www.LastIndexOf("/") + 1)       -destDir $tmp_dir -wc $net
doDownload -url $openSSL_www    -destFile $openSSL_www.Substring($openSSL_www.LastIndexOf("/") + 1)       -destDir $tmp_dir -wc $net
doDownload -url $boost_www      -destFile $boost_www.Substring($boost_www.LastIndexOf("/") + 1)           -destDir $tmp_dir -wc $net
doDownload -url $libovr_www     -destFile $libovr_www.Substring($libovr_www.LastIndexOf("/") + 1)         -destDir $tmp_dir -wc $net
doDownload -url $glew_www       -destFile $glew_www.Substring($glew_www.LastIndexOf("/") + 1)             -destDir $tmp_dir -wc $net
doDownload -url $opencv_www     -destFile $opencv_www.Substring($opencv_www.LastIndexOf("/") + 1)         -destDir $tmp_dir -wc $net

# only download Ogre if we're using VS 2012 or earlier
if ($vsVersion -eq "Visual Studio 10 2010" -Or $vsVersion -eq "Visual Studio 11 2012")
{
	doDownload -url $Ogre_SDK_www   -destFile $Ogre_SDK_www.Substring($Ogre_SDK_www.LastIndexOf("/") + 1)     -destDir $tmp_dir -wc $net
}

# check for existing code
foreach ($item in (dir $projectRoot)) {
    if ($item.Name.ToLower() -like "*sigviewer*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the SIGViewer source directory? (y/n)"
        if (!($conf.ToLower().StartsWith("y"))) {
            remove-item "$projectRoot\$($item.Name)" -recurse
        }
    }
    elseif ($item.Name.ToLower() -like "*sigservice*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the SIGService source directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("SIGSERVICE_ROOT_PATH", "$projectRoot\$($item.Name)")
        } else {
            remove-item "$projectRoot\$($item.Name)"
        }
    }
    elseif ($item.Name.ToLower() -like "*x3d*") {
        echo "Existing directory found: $($item.Name)"
        $conf = read-Host -prompt "Should this be used as-is as the X3D source directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("X3D_ROOT_PATH", "$projectRoot\$($item.Name)")
        } else {
            remove-item "$projectRoot\$($item.Name)"
        }
    }
}
foreach ($item in (dir $projectDepsRoot)) {
    if ($item.Name.ToLower() -like "*ogre*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the Ogre SDK directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("OGRE_SDK", "$projectDepsRoot\$($item.Name)")
        } else {
            remove-item "$projectDepsRoot\$($item.Name)"
        }
    }
    elseif ($item.Name.ToLower() -like "*opencv*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the OpenCV source directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("OPENCV_ROOT", "$projectDepsRoot\$($item.Name)")
        } else {
            remove-item "$projectDepsRoot\$($item.Name)"
        }
    }
    elseif ($item.Name.ToLower() -like "*cegui-[0-9]*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the CEGUI source directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("CEGUI_ROOT_PATH", "$projectDepsRoot\$($item.Name)")
        } else {
            remove-item "$projectDepsRoot\$($item.Name)"
        }
    }
    elseif ($item.Name.ToLower() -like "*cegui-deps*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the CEGUI DEPS directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("CEGUI_DEPS_ROOT", "$projectDepsRoot\$($item.Name)")
        } else {
            remove-item "$projectDepsRoot\$($item.Name)"
        }
    }
    elseif ($item.Name.ToLower() -like "*libssh2*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the libSSH2 directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("LIBSSH2_ROOT_PATH", "$projectDepsRoot\$($item.Name)")
        } else {
            remove-item "$projectDepsRoot\$($item.Name)"
        }
    }
    elseif ($item.Name.ToLower() -like "*openssl*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the OpenSSL directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("OPENSSL_ROOT_DIR", "$projectDepsRoot\$($item.Name)")
        } else {
            remove-item "$projectDepsRoot\$($item.Name)"
        }
    }
    elseif ($item.Name.ToLower() -like "*boost*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the Boost source directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("BOOST_ROOT", "$projectDepsRoot\$($item.Name)")
        } else {
            remove-item "$projectDepsRoot\$($item.Name)"
        }
    }
    elseif ($item.Name.ToLower() -like "*oculussdk*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the Oculus SDK directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("LIBOVR_ROOT_PATH", "$projectDepsRoot\$($item.Name)\LibOVR")
        } else {
            remove-item "$projectDepsRoot\$($item.Name)"
        }
    }
    elseif ($item.Name.ToLower() -like "*glew*") {
        echo "Existing directory found: $($item.Name)"
        $conf = Read-Host -prompt "Should this be used as-is as the GLEW root directory? (y/n)"
        if (($conf.ToLower().StartsWith("y"))) {
            $build_vars.Set_Item("GLEW_ROOT_PATH", "$projectDepsRoot\$($item.Name)")
        } else {
            remove-item "$projectDepsRoot\$($item.Name)"
        }
    }
}


# extract archives for any code not found above
if (!($build_vars.Get_Item("SIGSERVICE_ROOT_PATH"))) {
    iex "$7zX -o$projectRoot $tmp_dir\sigservice.zip"
    $build_vars.Set_Item("SIGSERVICE_ROOT_PATH", "$projectRoot\SIGService-master")
}

if (!($build_vars.Get_Item("X3D_ROOT_PATH"))) {
    iex "$7zX -o$projectRoot $tmp_dir\x3d.zip"
    $build_vars.Set_Item("X3D_ROOT_PATH", "$projectRoot\x3d-master")
}

if (!($build_vars.Get_Item("OGRE_SDK"))) {
    iex "$tmp_dir\$($Ogre_SDK_www.Substring($Ogre_SDK_www.LastIndexOf(""/"") + 1)) -o$projectDepsRoot -y"
    $build_vars.Set_Item("OGRE_SDK", "$projectDepsRoot\OgreSDK_vc11_v1-9-0")
}

if (!($build_vars.Get_Item("OPENCV_ROOT"))) {
    iex "$tmp_dir\$($opencv_www.Substring($opencv_www.LastIndexOf(""/"") +1)) -o$projectDepsRoot -y"
    $build_vars.Set_Item("OPENCV_ROOT", "$projectDepsRoot\opencv")
}

if (!($build_vars.Get_Item("CEGUI_ROOT_PATH"))) {
    iex "$7zX -o$projectDepsRoot $tmp_dir\$($CEGUI_www.Substring($CEGUI_www.LastIndexOf(""/"") + 1))"
    $build_vars.Set_Item("CEGUI_ROOT_PATH", "$projectDepsRoot\cegui-0.8.7")
}

if (!($build_vars.Get_Item("CEGUI_DEPS_ROOT"))) {
    iex "$7zX -o$projectDepsRoot $tmp_dir\$($CEGUI_deps_www.Substring($CEGUI_deps_www.LastIndexOf(""/"") + 1))"
    $build_vars.Set_Item("CEGUI_DEPS_ROOT", "$projectDepsRoot\cegui-deps-0.8.x-src")
}

if (!($build_vars.Get_Item("GLEW_ROOT_PATH"))) {
    iex "$7zX -o$projectDepsRoot $tmp_dir\$($glew_www.Substring($glew_www.LastIndexOf(""/"") + 1))"
    $build_vars.Set_Item("GLEW_ROOT_PATH", "$projectDepsRoot\glew-2.0.0")
}

if (!($build_vars.Get_Item("LIBSSH2_ROOT_PATH"))) {
    iex "$7zX -o$tmp_dir $tmp_dir\$($libSSH2_www.Substring($libSSH2_www.LastIndexOf(""/"") + 1))"
    iex "$7zX -o$projectDepsRoot $tmp_dir\libssh2-1.7.0.tar"
    $build_vars.Set_Item("LIBSSH2_ROOT_PATH", "$projectDepsRoot\libssh2-1.7.0")
}

if (!($build_vars.Get_Item("OPENSSL_ROOT_DIR"))) {
    $filename = "$($openSSL_www.Substring($openSSL_www.LastIndexOf(""/"") + 1))"
    iex "$7zX -o$projectDepsRoot\$($filename.substring(0, $filename.length-4)) $tmp_dir\$filename"
    $build_vars.Set_Item("OPENSSL_ROOT_DIR", "$projectDepsRoot\openssl-0.9.8k_WIN32")
}

if (!($build_vars.Get_Item("BOOST_ROOT"))) {
    iex "$7zX -o$projectDepsRoot $tmp_dir\$($boost_www.Substring($boost_www.LastIndexOf(""/"") + 1))"
    $build_vars.Set_Item("BOOST_ROOT", "$projectDepsRoot\boost_1_61_0")
}

if (!($build_vars.Get_Item("LIBOVR_ROOT_PATH"))) {
    iex "$7zX -o$projectDepsRoot $tmp_dir\$($libovr_www.Substring($libovr_www.LastIndexOf(""/"") + 1))"
    $build_vars.Set_Item("LIBOVR_ROOT_PATH", "$projectDepsRoot\OculusSDK\LibOVR")
}


echo ''
echo '=============================='
echo '     BUILD CONFIGURATION      '
echo '=============================='
echo "    Project Root:`t   $projectRoot"
echo "    CMake:       `t   $cmakePath"
foreach ($item in $build_vars.GetEnumerator()) {
    "{0, -3} {1, -22} {2, -80}" -f `
    "", $($item.Name + ":"), $item.Value
}
echo "    Visual Studio:"
echo "                  Version: $vsVersion"
echo "                  Tools:   $($build_vars.Get_Item(""VS_TOOLS_PATH""))"
echo ''
$t = Read-Host "Press Enter to continue..."

$dev_script = "$scriptPath\scripts\setenv.bat"

sc $dev_script '' -en ASCII
ac $dev_script '@echo off'
ac $dev_script 'title S I G V E R S E  x86 Release'
ac $dev_script 'rem --------------------------------------------------------'
ac $dev_script 'rem This is a script-generated file! Edit at your own risk!'
ac $dev_script 'rem --------------------------------------------------------'
ac $dev_script "set SIGVERSE_ROOT=""$projectRoot"""
ac $dev_script 'echo Checking for JDK path...'
ac $dev_script "call $scriptPath\scripts\find_jdk.bat"

ac $dev_script "set CMAKE=$cmake"
ac $dev_script "set VS_VERSION=""$vsVersion"""

foreach ($item in $build_vars.GetEnumerator()) {
    ac $dev_script "set $($item.Name)=""$($item.Value)"""
    ac $dev_script "if not exist %$($item.Name)% ("
    ac $dev_script "  echo $($item.Name) directory could not be found: %$($item.Name)%"
    ac $dev_script "  goto error"
    ac $dev_script ")"
}

# Include paths used by VS
ac $dev_script "set SIGBUILD_SIGSERVICE_INC=$($build_vars.Get_Item(""SIGSERVICE_ROOT_PATH""))\Windows\SIGService"
ac $dev_script "set SIGBUILD_X3D_INC=$($build_vars.Get_Item(""X3D_ROOT_PATH""))\parser\cpp\X3DParser"
ac $dev_script "set SIGBUILD_OGRE_INC=$($build_vars.Get_Item(""OGRE_SDK""))\include"
ac $dev_script "set SIGBUILD_BOOST_INC=$($build_vars.Get_Item(""BOOST_ROOT""))"
ac $dev_script "set SIGBUILD_CEGUI_INC=$($build_vars.Get_Item(""CEGUI_ROOT_PATH""))\cegui\include"
ac $dev_script "set SIGBUILD_LIBSSH2_INC=$($build_vars.Get_Item(""LIBSSH2_ROOT_PATH""))\include"
ac $dev_script "set SIGBUILD_OPENSSL_INC=$($build_vars.Get_Item(""OPENSSL_ROOT_DIR""))\include"
ac $dev_script "set SIGBUILD_LIBOVR_INC=$($build_vars.Get_Item(""LIBOVR_ROOT_PATH""))\Include;$($build_vars.Get_Item(""LIBOVR_ROOT_PATH""))\LibOVRKernel\Src"

# Lib paths used by VS
ac $dev_script "set SIGBUILD_SIGSERVICE_LIB=$($build_vars.Get_Item(""SIGSERVICE_ROOT_PATH""))\Windows\Release_2010"
ac $dev_script "set SIGBUILD_X3D_LIB=$($build_vars.Get_Item(""X3D_ROOT_PATH""))\parser\cpp\Release"
ac $dev_script "set SIGBUILD_OGRE_LIB=$($build_vars.Get_Item(""OGRE_SDK""))\lib"
ac $dev_script "set SIGBUILD_BOOST_LIB=$($build_vars.Get_Item(""BOOST_ROOT""))\stage\lib"
ac $dev_script "set SIGBUILD_CEGUI_LIB=$($build_vars.Get_Item(""CEGUI_ROOT_PATH""))\lib;$($build_vars.Get_Item(""CEGUI_ROOT_PATH""))\dependencies\lib\static"
ac $dev_script "set SIGBUILD_LIBSSH2_LIB=$($build_vars.Get_Item(""LIBSSH2_ROOT_PATH""))\build\src\Release"
ac $dev_script "set SIGBUILD_OPENSSL_LIB=$($build_vars.Get_Item(""OPENSSL_ROOT_DIR""))\lib"
ac $dev_script "set SIGBUILD_ZLIB_LIB=$($build_vars.Get_Item(""CEGUI_ROOT_PATH""))\dependencies\lib\static"
ac $dev_script "set SIGBUILD_LIBOVR_LIB=$($build_vars.Get_Item(""LIBOVR_ROOT_PATH""))\Lib\Windows\Win32\Release\VS2015"
ac $dev_script "set SIGBUILD_GLEW_LIB=$($build_vars.Get_Item(""GLEW_ROOT_PATH""))\lib\Release\Win32"

ac $dev_script "call %VS_TOOLS_PATH%\VsDevCmd.bat"

#ac $dev_script 'endlocal'
ac $dev_script 'if errorlevel 0 goto end'

ac $dev_script ':error'
ac $dev_script 'exit /B 1'

ac $dev_script ':end'

# make shortcut to setenv.bat
$WScriptShell = New-Object -ComObject WScript.Shell
$dev_script_shortcut = $WScriptShell.CreateShortcut("$projectRoot\SIGVerse_x86_Release.lnk")
$dev_script_shortcut.TargetPath = $env:ComSpec
$dev_script_shortcut.Arguments = "/K $dev_script"
$dev_script_shortcut.Save()

# Attempt to build all dependencies
cd "$scriptPath\scripts"
cmd /C 'build_boost.bat'
if ( $lastexitcode -ne 0) {
    echo 'ERROR: Boost build failed! Terminating build scripts'
    cd $projectRoot
    exit
}

cmd /C 'build_cegui.bat'
if ( $lastexitcode -ne 0) {
    echo 'ERROR: CEGUI build failed! Terminating build scripts'
    cd $projectRoot
    exit
}

cmd /C 'build_libssh2.bat'
if ( $lastexitcode -ne 0) {
    echo 'ERROR: LIBSSH2 build failed! Terminating build scripts'
    cd $projectRoot
    exit
}

cmd /C 'build_x3d.bat'
if ( $lastexitcode -ne 0) {
    echo 'ERROR: X3D build failed! Terminating build scripts'
    cd $projectRoot
    exit
}

cmd /C 'build_sigservice.bat'
if ( $lastexitcode -ne 0) {
    echo 'ERROR: SIGService build failed! Terminating build scripts'
    cd $projectRoot
    exit
}

cmd /C 'build_opencv.bat'
if ( $lastexitcode -ne 0) {
    echo 'ERROR: OpenCV build failed! Terminating build scripts'
    cd $projectRoot
    exit
}

echo ""
echo " ======================= "
echo " SIGVerse SETUP COMPLETE"
echo " ======================= "
echo ""
cd $projectRoot
